### ISSUE TITLE
Incorrect result when running the same xpath against a similar document

### ISSUE THREAD

**Issue by ymor** [FontoXML/fontoxpath#222]
See below example test. 

```javascript
const nameSpaceDocument = new slimdom.Document();
const namespaceElement = nameSpaceDocument.createElementNS("http://test.com", "name");
namespaceElement.textContent = 'name1'
nameSpaceDocument.appendChild(namespaceElement);

chai.assert.equal(evaluateXPathToString('/name/text()',	nameSpaceDocument), 'name1')

const slimdomDocument = new slimdom.Document();
const nameElement = slimdomDocument.createElement("name");
nameElement.textContent = 'name2'
slimdomDocument.appendChild(nameElement);

chai.assert.equal(evaluateXPathToString('/name/text()',	slimdomDocument), 'name2')
```

The first document includes a namespace on the name element. The second document does not.
The second assertion will fail, however if you change the order of the assertions it works as expected. If you change the second xpath to just be '/name' it also works as expected.
Looks to be a caching issue?

**Comment by ymor** [comment]
I've just noticed there is a disable cache option.  Specifying this makes it work as expected.

**Comment by DrRataplan** [comment]
Hi there!

Good find! Indeed, seems like a caching issue. I'll take a look at it.

Btw, personally, I prefer to use the `namespaceResolver` option. I think it is bad practice to 'assume' the prefix -> namespace bindings in your document match the prefix -> namespace bindings in the XPath. The 'default' is handy for quickly testing something, but using it for production, I would not depend on that. It may be that your first document binds another prefix to the `'http://test.com'` uri; causing the expression to not work anymore. Remember that prefixes are just visual and have no real meaning in XML/XPath/XQuery. It is all about the namespace uri.

Nonetheless, this is a bug. Thanks for filing it!

**Comment by DrRataplan** [comment]
Found it, when compiling an expression, we try to return a fully cached expression if both the expression string the static contexts match another run. This fails here. We then try to get a 'half compiled' result, which can occur when someone just asked for a bucket of a selector, but did not execute it. Getting the bucket of a selector does not require static compilation, but does require a 'normal' compilation. However, this mistakingly gets _any_ compilation result. See https://github.com/FontoXML/fontoxpath/blob/master/src/parsing/compiledExpressionCache.ts#L129, which in the end does https://github.com/FontoXML/fontoxpath/blob/master/src/parsing/compiledExpressionCache.ts#L46.

The fix would be to _only_ use the 'halfcompiled' cache here.

**Comment by DrRataplan** [comment]
I just published 3.11.1, which fixes this issue. Could you verify this?

**Comment by ymor** [comment]
Hi,
Yes it is now working as expected. Thanks for the quick turnaround!
